feat: add `MovieInfo` screen and helper components

abhishek-gkmit opened the pull request:
- add `MovieCard` component
- add `TextWithIcon` component
- add types for components

Edit: The reason for the large file count is `fonts`

abhishek-gkmit committed:
feat: add `MovieInfo` screen and helper components
- add `MovieCard` component
- add `TextWithIcon` component
- add types for components

adiTrivsme commented on src/components/movieCard/MovieCard.tsx:
```diff
@@ -0,0 +1,29 @@
+import { StyleSheet, View, Text, TouchableOpacity, Image } from 'react-native';
+
+import styles from '@components/movieCard/style';
+import { memo } from 'react';
```
keep imports in order

"
basic imports - react, 3rd party lib, etc

components

assets - globalStyles, utils, services

./style.ts

"

Also you can import neighboring files directly - as in here `./style`

adiTrivsme commented on src/components/movieCard/MovieCard.tsx:
```diff
@@ -0,0 +1,29 @@
+        style={styles.imageStyle}
+        resizeMode="cover"
+      />
+        <Text numberOfLines={1} ellipsizeMode='tail' style={StyleSheet.compose(styles.headingStyle, headingStyle)}>
```
why using `Stylesheet.compose`??

abhishek-gkmit replied:
`Stylesheet.compose` is more efficient compared to the regular `array` method. Every time we use the regular `array` method to pass `styles` to components it creates a new array even if one of the `style` object is `null` or `falsy` but if we use the `Stylesheet.compose` method it will directly return the non-`falsy` or non-`null` style object without creating a new style object.

adiTrivsme replied:
Please provide the reference for the same on the group

adiTrivsme commented on src/components/movieCard/index.tsx:
```diff
@@ -0,0 +1,2 @@
+import MovieCard from '@components/movieCard/MovieCard';
```
same as commented above

adiTrivsme commented on src/components/movieCard/style.tsx:
```diff
@@ -0,0 +1,28 @@
+import colors from '@constants/colors';
+import { StyleSheet } from 'react-native';
```
fix import order

adiTrivsme commented on src/components/textWithIcon/TextWithIcon.tsx:
```diff
@@ -0,0 +1,27 @@
+        size={icon.size || 14}
+        color={icon.color ? icon.color : colors.primary}
+      />
+      <Text style={StyleSheet.compose(styles.textStyle, textStyle)}>
```
same as the above comment

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,277 @@
+import TextWithIcon from '@components/textWithIcon';
+import MovieCard from '@components/movieCard';
+import { ThemeAndStorageContext } from '@contexts/ThemeAndStorageContext';
+import colors from '@constants/colors';
```
keep imports sorted in order

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,277 @@
+  }, [page]);
+
+  useEffect(() => {
+    console.log('inital loading>>>>>>>>>');
```
remove logs

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,277 @@
+  useEffect(() => {
+    console.log('inital loading>>>>>>>>>');
+    (async () => {
+      const movieInfo = await getMovieInfo(params.movieId);
```
where are you handing loaders?

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,277 @@
+      <TouchableOpacity
+        activeOpacity={0.5}
+        style={styles.favouriteBtn}
+        onPress={() =>
```
extract to a function

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,277 @@
+    setIsFavourite(isFavourite);
+  }, [favouriteMovies, watchlist, setInWatchlist, setIsFavourite, movieInfo]);
+
+  return (
```
break this in small render components

refer - [COC](https://docs.google.com/document/d/1jVt9hZ0ydQaNt2b5hYdgpuFfd9PSOnZhK9tLSj9xxtI/edit?tab=t.0#heading=h.mvjz6n2qebkp)

adiTrivsme replied:
This still can be broken into smaller components, its fine for now but make sure to follow this in future

adiTrivsme commented on src/screens/movieInfo/styles.ts:
```diff
@@ -0,0 +1,148 @@
+  },
+
+  movieName: {
+    fontFamily: 'NunitoSans-Regular',
```
define constants for font related properties - family, size, etc

abhishek-gkmit committed:
fix: add suggested changes
- add font support and font constants for reusablity
- add `customLoader` component to be used accros project
- add `iconButton` component

abhishek-gkmit committed:
fix: add remaining types for `MovieInfo` screen

adiTrivsme commented on ios/MoviesExplorer/Info.plist:
```diff
@@ -1,82 +1,98 @@
+	</array>
+	<key>UIAppFonts</key>
+	<array>
+		<string>AntDesign.ttf</string>
```
do you really need all these icons?

adiTrivsme commented on src/components/customLoader/Loader.tsx:
```diff
@@ -0,0 +1,19 @@
+  return (
+    <View styles={styles.loaderContainer}>
+      <ActivityIndicator
+        size={isSmall ? 'small' : 'large'}
```
can directly take in the prop - `small || large`

adiTrivsme commented on src/components/customLoader/Loader.tsx:
```diff
@@ -0,0 +1,19 @@
+
+import colors from '@constants/colors';
+
+import styles from '@components/customLoader/styles.ts';
```
style can be imported directly here -  `./style`

adiTrivsme commented on src/components/customLoader/index.tsx:
```diff
@@ -0,0 +1,2 @@
+import Loader from '@components/customLoader/Loader';
```
same here - I have already mentioned this in the previous review, neighboring files can be imported using normal import statements

adiTrivsme commented on src/components/customLoader/styles.ts:
```diff
@@ -0,0 +1,12 @@
+import { StyleSheet } from 'react-native';
+
+import colors from '@constants/colors';
```
color should be placed in `theme`

abhishek-gkmit replied:
I will refactor this when I implement the `theme` feature

adiTrivsme commented on src/components/customLoader/styles.ts:
```diff
@@ -0,0 +1,12 @@
+import colors from '@constants/colors';
+
+const styles = StyleSheet.create({
+  loaderContainer: {
```
why have you imported `color` if not used

adiTrivsme commented on src/components/iconButton/IconButton.tsx:
```diff
@@ -0,0 +1,17 @@
+import { StyleSheet, Text, TouchableOpacity } from 'react-native';
+import MaterialCommunityIcon from 'react-native-vector-icons/MaterialCommunityIcons';
+
+import styles from '@components/iconButton/styles';
```
same as above

adiTrivsme commented on src/components/iconButton/IconButton.tsx:
```diff
@@ -0,0 +1,17 @@
+function IconButton({ icon, btnStyle, ...props }: IconButtonComponentProps) {
+  return (
+    <TouchableOpacity
+      style={StyleSheet.compose(styles.iconButton, btnStyle)}
```
use `[]` here

adiTrivsme commented on src/components/iconButton/index.ts:
```diff
@@ -0,0 +1,3 @@
+import IconButton from '@components/iconButton/IconButton';
```
same as above

adiTrivsme commented on src/components/movieCard/MovieCard.tsx:
fix issues based on above mentioned comments

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,301 @@
+  // memoized values
+  const isFavourite = useMemo(() => {
+    const isFavourite = favouriteMovies.some(
+      movieId => movieId === movieInfo.id + '',
```
why `movieInfo.id` not in dependency array?

adiTrivsme replied:
this has not been addressed yet

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,301 @@
+      isFavourite
+        ? removeFromFavouritesWrapper(movieInfo.id + '')
+        : addToFavouritesWrapper(movieInfo.id + '');
+  }, [isFavourite]);
```
same here

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,301 @@
+  }, []);
+
+  if (loading) {
+    return <Loader animating={true} />;
```
if using `animating` property why put this in condition?

abhishek-gkmit replied:
If `loading` is `true` then this component will only return the `Loader` else it will return what we actually want it to render

adiTrivsme replied:
My point is why not use the animating property directly, if you are using `animating` use it for the show-hide purpose directly

<img width="606" alt="Screenshot 2024-10-29 at 9 08 14 PM" src="https://github.com/user-attachments/assets/a5219c21-1cf1-423d-9c29-bf9d9906b79b">

adiTrivsme commented on src/screens/movieInfo/styles.ts:
```diff
@@ -0,0 +1,122 @@
+import colors from '@constants/colors';
+import { StyleSheet } from 'react-native';
```
import order

abhishek-gkmit committed:
fix: add suggested changes
- removed unnecessary font files from `android` and `ios`

adiTrivsme commented on src/components/iconButton/IconButton.tsx:
```diff
@@ -0,0 +1,17 @@
+function IconButton({ icon, btnStyle, ...props }: IconButtonComponentProps) {
+  return (
+    <TouchableOpacity
+      style={[styles.iconButton, btnStyle]}
```
why not use `Stylesheet.compose` here?

adiTrivsme commented on src/components/textWithIcon/TextWithIcon.tsx:
```diff
@@ -0,0 +1,26 @@
+import { View, Text } from 'react-native';
+import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
+
+import colors from '@constants/colors';
```
move this `color` to theme, why is this not resolved yet

abhishek-gkmit replied:
I will refactor this when I implement the `theme` feature

adiTrivsme commented on src/screens/movieInfo/MovieInfo.tsx:
```diff
@@ -0,0 +1,294 @@
+
+  const inWatchlist = useMemo(() => {
+    const inWatchlist = watchlist.some(
+      movieId => movieId === movieInfo.id + '',
```
same as above comment

abhishek-gkmit committed:
fix: add `useMemo` dependency and remove `animating` from `Loader`

abhishek-gkmit committed:
Merge branch 'dev' into feat/add-movie-info-screen
